### ISSUE TITLE
How to get original document whilst iterating through query results?

### ISSUE THREAD

**Issue by SaadAttieh** [nlfiedler/mokuroku#9]
I couldn't see an example of how to do this so I guessed it would just be calling `dbase.get` on the `doc_id` in the query result.
```
//adapted from tags example
    let results = dbase.query_by_key("tags", b"cat").unwrap();
    for result in results {
        let doc_id = str::from_utf8(&result.doc_id).unwrap().to_owned();
        println!("query result key: {:}", doc_id);
        println!("tag value: {}", str::from_utf8(&result.value).unwrap().to_owned());
        println!("get actual value  {:?}", dbase.get::<Asset,_>(result.doc_id).ok().unwrap());
    }

```

But I get the mutable borrow error:
```
error[E0502]: cannot borrow `dbase` as immutable because it is also borrowed as mutable
   --> src/main.rs:123:44
    |
118 |     let results = dbase.query_by_key("tags", b"cat").unwrap();
    |                   ----- mutable borrow occurs here
119 |     for result in results {
    |                   ------- mutable borrow later used here
...
123 |         println!("get actual value  {:?}", dbase.get::<Asset,_>(result.doc_id).ok().unwrap());
    |                                            ^^^^^ immutable borrow occurs here

```

**Comment by nlfiedler** [comment]
You are quite right that this use case was not considered when designing the API. For my purposes, I only needed the index values, and querying the full documents would come later as needed. It would require a different kind of iterator that returns the primary value along with the index values.

**Comment by nlfiedler** [comment]
Six years later I'm going to close this one out. Claude offers this analysis as an explanation:

> Your performance concern is valid — and it's the same one CouchDB has
>
> You're right to be wary, and the LSM-vs-B-tree distinction sharpens it rather than softening it:
>
> - A plain query() is a sequential range scan over one column family. That's the best case for any storage engine —
> block-cache friendly, prefetchable, low amplification.
> - include_docs turns each row into a random point lookup into a different CF. So a K-row result goes from "1
> sequential scan" to "1 scan + K random reads." That's the classic index→fetch random-I/O problem, and it scales
> linearly with result size.
> - On an LSM tree a point get can touch the memtable plus multiple SST levels (read amplification), though bloom
> filters knock most of that out. So per-lookup it's usually fine; the cost is the count of random lookups, not any
> single one.
>
> CouchDB has exactly this issue despite being a B-tree, and its own docs warn about it — they recommend emitting the
> data you need as the view value instead of using include_docs when performance matters. There's also the consistency
> caveat: the doc you fetch may be newer than the index entry that pointed at it (mokuroku already tracks staleness of
> index entries via sequence numbers, but the fetched doc is whatever's current).
